src/builders/thermodynamic.jl

```diff
@@ -0,0 +1,252 @@
+    mets = collect(mets)
+    stoi = A.stoichiometry(model)
+
+    # create thermodynamic variables
+    m = C.ConstraintTree(
+        :max_min_driving_force^C.variable() +
+        :log_metabolite_concentrations^C.variables(
+            keys = Symbol.(mets),
+            bounds = zip(log(concentration_lb), log(concentration_ub)),
+        ) +
+        :delta_G_reactions^C.variables(keys = Symbol.(rxns)),
+    )
+
+    #=
+    Build gibbs free energy relations
+    ΔG_rxn == ΔG0 + R * T * sum ( log_concentration_variable * stoichiometry_value )
+    =#
+    model_rxns = A.reactions(model)
+    model_mets = A.metabolites(model)
+    dG0s_met_ids_stoichs = Vector{Tuple{Float64,Vector{String},Vector{Float64}}}()
+    for rxn in rxns # prepare to create CT below
+        met_idxs, stoich_coeffs = findnz(stoi[:, findfirst(==(rxn), model_rxns)])
+        met_ids = model_mets[met_idxs]
+        dG0 = reaction_standard_gibbs_free_energies[rxn]
+        push!(dG0s_met_ids_stoichs, (dG0, met_ids, stoich_coeffs))
+    end
+
+    m *=
+        :delta_G_reaction_equations^C.ConstraintTree(
+            Symbol(rxn) => C.Constraint(
+                value = -m.delta_G_reactions[Symbol(rxn)].value +
+                        dG0 +
+                        R *
+                        T *
+                        sum(
+                            m.log_metabolite_concentrations[Symbol(met_id)].value * stoich for (met_id, stoich) in zip(met_ids, stoich_coeffs)
+                        ),
+                bound = 0.0,
+            ) for (rxn, (dG0, met_ids, stoich_coeffs)) in zip(rxns, dG0s_met_ids_stoichs)
+        )
+
+    #=
+    Set proton log concentration to zero so that it won't impact any
+    calculations (biothermodynamics assumption). Also set water concentrations
+    to zero (aqueous condition assumptions). How true is "aqueous conditions"?
+    Debatable...
+    =#
+    for met in [Symbol.(proton_ids); Symbol.(water_ids)]
+        haskey(m.log_metabolite_concentrations, met) &&
+            (m.log_metabolite_concentrations[met].bound = 0.0)
+    end
+
+    #=
+    Add thermodynamic feasibility constraint (ΔG < 0 for a feasible reaction in flux direction).
+    Add objective constraint to solve max min problem.
+    =#
+    m *=
+        :reaction_delta_G_margin^C.ConstraintTree(
+            Symbol(rxn) => C.Constraint(
+                value = m.delta_G_reactions[Symbol(rxn)].value *
+                        sign(get(reference_flux, rxn, 1.0)),
+                bound = (-Inf, 0.0),
+            ) for rxn in rxns
+        )
+
+    m *=
+        :min_driving_force_margin^C.ConstraintTree(
+            Symbol(rxn) => C.Constraint(
+                value = m.max_min_driving_force.value +
+                        m.delta_G_reactions[Symbol(rxn)].value *
+                        sign(get(reference_flux, rxn, 1.0)),
+                bound = (-Inf, 0.0),
+            ) for rxn in rxns
+        )
```

Comment:
merge all `*=` into a single `*`, no need to do `*=` when we return the result immediately

Reply:
okay, but right now it reads nicely, with each block separated from the others